packages/apputils-extension/src/index.ts

```diff
@@ -218,31 +212,106 @@ const state: JupyterLabPlugin<IStateDB> = {
   id: '@jupyterlab/apputils-extension:state',
   autoStart: true,
   provides: IStateDB,
-  activate: (app: JupyterLab) => {
+  requires: [IRouter],
+  activate: (app: JupyterLab, router: IRouter) => {
+    let command: string;
+    let debouncer: number;
+    let resolved = false;
+    let workspace = '';
+
+    const { commands, info, serviceManager } = app;
+    const { workspaces } = serviceManager;
+    const transform = new PromiseDelegate<StateDB.DataTransform>();
     const state = new StateDB({
-      namespace: app.info.namespace,
-      when: app.restored.then(() => { /* no-op */ })
+      namespace: info.namespace,
+      transform: transform.promise
     });
-    const version = app.info.version;
-    const key = 'statedb:version';
-    const fetch = state.fetch(key);
-    const save = () => state.save(key, { version });
-    const reset = () => state.clear().then(save);
-    const check = (value: JSONObject) => {
-      let old = value && value['version'];
-      if (!old || old !== version) {
-        const previous = old || 'unknown';
-        console.log(`Upgraded: ${previous} to ${version}; Resetting DB.`);
-        return reset();
+    const disposables = new DisposableSet();
+    const pattern = /^\/workspaces\/(.+)/;
+    const unload = () => {
+      disposables.dispose();
+      router.routed.disconnect(unload, state);
+
+      // If the request that was routed did not contain a workspace,
+      // leave the database intact.
+      if (!resolved) {
+        transform.resolve({ type: 'cancel', contents: null });
       }
     };
 
-    app.commands.addCommand(CommandIDs.clearStateDB, {
+    command = CommandIDs.clearState;
+    commands.addCommand(command, {
       label: 'Clear Application Restore State',
       execute: () => state.clear()
     });
 
-    return fetch.then(check, reset).then(() => state);
+    command = CommandIDs.saveState;
+    commands.addCommand(command, {
+      label: () => `Save Workspace (${workspace})`,
+      isEnabled: () => !!workspace,
+      execute: () => {
+        if (!workspace) {
+          return;
+        }
+
+        const id = workspace;
+        const metadata = { id };
+
+        if (debouncer) {
+          window.clearTimeout(debouncer);
+        }
+
+        debouncer = window.setTimeout(() => {
+          state.toJSON()
+            .then(data => workspaces.save(id, { data, metadata }))
+            .catch(reason => {
+              console.warn(`Saving workspace (${id}) failed.`, reason);
+            });
+        }, 2000);
```

Comment:
Let's use a module const for this and add documentation so our future selves can understand the meaning of `2000`.

Reply:
👍